# Arhat Shah Assignment 11

In [2]:
#imports

import pandas as pd
from sklearn.datasets import load_iris
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC

In [4]:
# Perform any necessary data preparation steps and EDA.

# load the iris dataset
iris = load_iris()

# create a dataframe with the data and target
df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])
df['target'] = iris['target']

# check for missing values and data types
df.info()

# split the data into features and target
X = df.drop('target', axis=1)
y = df['target']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


In [5]:
# Normalize the numeric features and encode any categorical features.

# normalize the numeric features
scaler = StandardScaler()
X[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']] = scaler.fit_transform(X[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']])

# encode the target
encoder = LabelEncoder()
y = encoder.fit_transform(y)

In [7]:
# Partition the data into train/test sets.

# split the data into train/test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
# Build the model that predicts the target variable, and use GridSearchCV to tune at least two hyperparameters, 
# and perform cross validation.  You'll need to read the documentation of the SVM algorithm to identify the 
# hyperparameters.

# create a SVM object
svm = SVC()

# define the hyperparameter grid
param_grid = {'kernel': ['linear', 'rbf'], 'C': [0.1, 1, 10], 'gamma': ['scale', 'auto']}

# create a GridSearchCV object
grid_search = GridSearchCV(estimator=svm, param_grid=param_grid, cv=5, n_jobs=-1)

# fit the GridSearchCV object to the training data
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=SVC(), n_jobs=-1,
             param_grid={'C': [0.1, 1, 10], 'gamma': ['scale', 'auto'],
                         'kernel': ['linear', 'rbf']})

In [10]:
# Display: 1) the best model/parameters and 2) the best score that was obtained.

# print the best model/parameters and the best score
print(grid_search.best_estimator_)
print(grid_search.best_params_)
print(grid_search.best_score_)

SVC(C=10, kernel='linear')
{'C': 10, 'gamma': 'scale', 'kernel': 'linear'}
0.9583333333333334


In [11]:
# How do you feel about the model? Did it accurately predict the target? 
# Evaluate the scores from GridSearchCV for the training set and the test set to diagnose any bias-variance problems.

# evaluate the model on the training set and the test set
print('Training accuracy:', grid_search.score(X_train, y_train))
print('Test accuracy:', grid_search.score(X_test, y_test))

Training accuracy: 0.9666666666666667
Test accuracy: 0.9666666666666667


In general, the iris dataset showed excellent performance by the SVM model. By using the GridSearchCV algorithm, we were able to discover the most suitable hyperparameters for the SVM model, resulting in a highly accurate model. The accuracy of the training and test data was both 0.967, indicating that the model is not excessively fitted to the training data and can effectively handle new data. As a result, I am confident in the model's capacity to make precise predictions for the target variable.



